esphome/components/thermostat/thermostat_climate.h

```diff
@@ -8,6 +8,28 @@
 namespace esphome {
 namespace thermostat {
 
+enum ThermostatClimateTimerIndex : size_t {
+  TIMER_COOLING_MAX_RUN_TIME = 0,
+  TIMER_COOLING_OFF = 1,
+  TIMER_COOLING_ON = 2,
+  TIMER_FANNING_OFF = 3,
+  TIMER_FANNING_ON = 4,
+  TIMER_HEATING_MAX_RUN_TIME = 5,
+  TIMER_HEATING_OFF = 6,
+  TIMER_HEATING_ON = 7,
+  TIMER_IDLE_ON = 8,
+  TIMER_TIMER_COUNT = 9
+};
+
+class ThermostatClimate;  // ...so we can use it in the struct below
+
+struct ThermostatClimateTimer {
+  const std::string name;
+  bool active;
+  uint32_t time;
+  void (esphome::thermostat::ThermostatClimate::*func)();
```

Comment:
```suggestion
  void (ThermostatClimate::*func)();
```

Not a huge fan of the use of raw function pointers, but I see how in this case it would be a bit difficult otherwise (`timers_` array would have to be created at runtime)

Reply:
Good catch -- I completely agree. That said, I wasn't quite sure how best to implement this in a way that didn't result in a bunch of duplicated code. For what it's worth, these get run through `bind` [here](https://github.com/esphome/esphome/pull/2114/files#diff-1d51f5af332e771ee2453ab457675345b9ff096b9215a5a397ec80a20e97e4e1R649) which _I think_ makes them a bit safer...maybe? I'm new to this specific architecture and totally open to suggestions if something better comes to mind. 😄

Reply:
I think something like this might work when the type here is changed to `std::function<void()> func`:

```cpp
// run this in setup somewhere
timer_ = {
  {"cool_run", false, 0, [this] { this->cooling_max_run_time_timer_callback() }},
  {"cool_run", false, 0, [this] {
    // or if body short just put the code directly here
  }},
  // ...
}
```

Reply:
Ah sweet! I will give it a try. 😄 Thank you!

Reply:
You don't even need a lambda, `std::bind(&ThermostatClimate::cooling_max_run_time_timer_callback_, this)` should work as well. (Okay, I guess this technically still uses raw function pointers, but given their immediate conversion I don't think any of the downsides apply here.)

Reply:
Hmm, seems to keep getting hung up on the function pointers being lvalues while `std::function` expects an rvalue. I'll tinker more this evening -- was just trying to see if I could hammer it out during lunch. 😆

Reply:
Figured it out 😁